### ISSUE TITLE
Cargo incorrectly identifies transitive dependency versions when compiling new dependencies

### ISSUE THREAD

**Issue by CodingAnarchy** [rust-lang/cargo#14059]
### Problem

When adding a new dependency to `Cargo.toml`, if a transitive dependency matches the same name as a transitive dependency of some other crate but has a different version, Cargo fails to obtain the new version and compile with it. This can lead to compilation errors as a result of the changes between the versions.

### Steps

1. Start with a `Cargo.toml` file that depends on `moka = { version = "= 0.12.5", features = ["sync", "future"] }`. This will require, as a transitive second-order dependency, `concurrent-queue = "2.4.0"`.
2. Compile the build to generate the `Cargo.lock` file.
3. Add a requirement like `chewdata = "= 2.11.0"` which has a dependency for `async-channel = "2.3"`. This will require `concurrent-queue = "2.5"` in the background as a transitive second-order dependency.
4. Attempt to compile again, and the following error results:

```
error[E0432]: unresolved import `concurrent_queue::ForcePushError`
  --> ~/.cargo/registry/src/index.crates.io-6f17d22bba15001f/async-channel-2.3.0/src/lib.rs:50:41
   |
50 | use concurrent_queue::{ConcurrentQueue, ForcePushError, PopError, PushError};
   |                                         ^^^^^^^^^^^^^^ no `ForcePushError` in the root

error[E0599]: no method named `force_push` found for struct `ConcurrentQueue` in the current scope
   --> ~/.cargo/registry/src/index.crates.io-6f17d22bba15001f/async-channel-2.3.0/src/lib.rs:313:34
    |
313 |         match self.channel.queue.force_push(msg) {
    |                                  ^^^^^^^^^^ method not found in `ConcurrentQueue<T>`

Some errors have detailed explanations: E0432, E0599.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `async-channel` (lib) due to 2 previous errors
```

This is apparently a result of incorrectly using `concurrent-queue 2.4.x` for compiling `async-channel`, which lacks the error types being required for use in this code.

### Possible Solution(s)

It seems like the logic that Cargo is doing to download and resolve these second-order transitive dependencies does not take into account the version information properly in some cases. I'm not sure how best to fix that, though.

### Notes

Specifying `concurrent-queue = "2.5"` as a top-level dependency in the application resolves this error. Only when this transitive dependency is added without that top-level version requirement is the error present.

### Version

```text
cargo 1.77.2 (e52e36006 2024-03-26)
release: 1.77.2
commit-hash: e52e360061cacbbeac79f7f1215a7a90b6f08442
commit-date: 2024-03-26
host: aarch64-apple-darwin
libgit2: 1.7.2 (sys:0.18.2 vendored)
libcurl: 8.6.0 (sys:0.4.70+curl-8.5.0 system ssl:(SecureTransport) LibreSSL/3.3.6)
ssl: OpenSSL 1.1.1w  11 Sep 2023
os: Mac OS 14.5.0 [64-bit]
```

**Comment by epage** [comment]
Tried to simplify the reproduction steps
```console
$ cargo new cargo-14059
$ cd cargo-14059
$ cargo add moka@=0.12.5 -Fsync,future
$ cargo check
$ git add . && git commit -m "Initial"
$ cargo add chewdata@=2.11.0
$ cargo check
```

However, this did not reproduce the problem with Cargo 1.79.

**Comment by epage** [comment]
```console
$ git checkout HEAD~
$ rm Cargo.lock
$ cargo +1.77 check
$ git diff
```
and the lockfile was unchanged
```console
$ git checkout main
$ git checkout HEAD~ -- Cargo.lock
$ cargo +1.77 check
$ git add .
$ git status
```
and the build worked and the lockfile was unchanged

**Comment by CodingAnarchy** [comment]
@epage I looked further into this, and it seems the issue was with `async-channel 2.3.0` having the wrong dependency bounds. This was fixed in `2.3.1`, and I think we just hit a weird edge case when that change was made relative to our existing lockfile. Trying again with the more limited case in a fresh project also did not replicate for me.

My apologies for the noise on this. I'll go ahead and close this as resolved in the original project.
